Plutarch/List.hs

```diff
@@ -0,0 +1,189 @@
+ > instance PConstant a b => PConstant [a] (PList b)
+-}
+module Plutarch.List (
+  PList (..),
+
+  -- * Query
+  pelem,
+  plength,
+  pnull,
+
+  -- * Construction
+  pnil,
+  psingleton,
+
+  -- * Combine
+  pconcat,
+
+  -- * Traversals
+  pmap,
+  pfilter,
+
+  -- * Catamorphisms
+  precList,
+  punsafeHead,
+  punsafeTail,
+  pfoldr,
+
+  -- * Casting
+  pforgetList,
+) where
+
+import Plutarch
+import Plutarch.Bool
+import Plutarch.Builtin
+import Plutarch.Integer
+import Plutarch.Prelude
+import qualified PlutusCore as PLC
+
+--------------------------------------------------------------------------------
+
+data PList (a :: k -> Type) (s :: k)
+  = PCons (Term s a) (Term s (PList a))
+  | PNil
+
+-- Instances
+
+instance PIsData a => PlutusType (PList a) where
+  type PInner (PList a) b = PBuiltinList (PAsData a)
+  pcon' :: forall s. PList a s -> forall b. Term s (PInner (PList a) b)
+  pcon' (PCons x xs) = pconsBuiltin # (pdata x) # (pto xs)
+  pcon' PNil = pemptyList
+  pmatch' xs f =
+    pforce $ pif (pnullBuiltin # xs) (pdelay (f PNil)) (pdelay (f (PCons (pfromData $ pheadBuiltin # xs) (punsafeFrom $ ptailBuiltin # xs))))
+
+instance PEq a => PEq (PList a) where
+  (#==) xs ys =
+    punsafeBuiltin PLC.EqualsData
+      # (pmkList # punsafeCoerce xs)
+      # (pmkList # punsafeCoerce ys)
+
+--------------------------------------------------------------------------------
+-- Construction
+
+-- | Create a singleton list from an element
+psingleton :: PIsData a => Term s (a :--> PList a)
+psingleton = plam $ \x -> pcon (PCons x (pcon PNil))
+
+-- | The empty list
+pnil :: PIsData a => Term s (PList a)
+pnil = pcon PNil
+
+--------------------------------------------------------------------------------
+-- Querying
+
+-- | / O(n) /. Check if element is in the list
+pelem :: (PIsData a, PEq a) => Term s (a :--> PList a :--> PBool)
+pelem =
+  phoistAcyclic $
+    plam $ \needle ->
+      precList
+        (\self x xs -> pif (x #== needle) (pcon PTrue) (self # xs))
+        (\_self -> pcon PFalse)
+
+-- | / O(n) /. Count the number of elements in the list
+plength :: PIsData a => Term s (PList a :--> PInteger)
+plength =
+  plet
+    ( pfix #$ plam $ \self ls n ->
+        pmatch ls $ \case
+          PCons _x xs -> (self # xs # n + 1)
+          PNil -> n
+    )
+    $ \go -> plam $ \xs -> go # xs # 0
+
+pnull :: PIsData a => Term s (PList a :--> PBool)
+pnull =
+  plam $ \xs ->
+    pmatch xs $ \case
+      PCons _ _ -> pcon PFalse
+      PNil -> pcon PTrue
+    
+
+--------------------------------------------------------------------------------
+
+-- | Meta recursive list eliminator
+precList :: PIsData a => (Term s (PList a :--> r) -> Term s a -> Term s (PList a) -> Term s r) -> (Term s (PList a :--> r) -> Term s r) -> Term s (PList a :--> r)
+precList mcons mnil =
+  pfix #$ plam $ \self xs ->
+    pmatch xs $ \case
+      PCons x xs -> mcons self x xs
+      PNil -> mnil self
+
```

Comment:
I wonder what the budget difference here would be when using [`ChooseList`](https://github.com/input-output-hk/plutus/blob/c8c5183f7facd967d48fe07b3b14465b8dd48fe7/plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs#L286-L297) directly.

Reply:
Yeah, good point. Maybe using `ChooseList` is better for the `pmatch'` implementation. I'll figure out the budget difference and pick the best one.

Reply:
Alright, there's a slight benefit to be gained by using `ChooseList`:

```hs
-- pif + pnullBuiltin
Right (ExBudget {exBudgetCPU = ExCPU 895561, exBudgetMemory = ExMemory 2165},[],<Script>)
-- pchooseListBuiltin
Right (ExBudget {exBudgetCPU = ExCPU 776468, exBudgetMemory = ExMemory 1764},[],<Script>)
```